Pooling allocator: add a reuse-affinity policy.

cfallin opened the pull request:
(Builds on #3697.)

This policy attempts to reuse the same instance slot for subsequent
instantiations of the same module. This is particularly useful when
using a pooling backend such as memfd that benefits from this reuse: for
example, in the memfd case, instantiating the same module into the same
slot allows us to avoid several calls to mmap() because the same
mappings can be reused.

The policy tracks a freelist per "compiled module ID", and when
allocating a slot for an instance, tries these two options in order:

1. A slot from the freelist for this module (i.e., last used for another
   instantiation of this particular module), or
2. A slot that was last used by some other module or never before.

The "victim" slot for choice 2 is randomly chosen.

The data structures are carefully designed so that all updates are O(1),
and there is no retry-loop in any of the random selection.

This policy is now the default when the memfd backend is selected via
the `memfd-allocator` feature flag.

cfallin commented:
After writing this, it occurs to me that the reuse policy as stated will choose with equal probability a *module* whose freelist we still from, but this does not imply equal probability for any *slot* to be stolen.

In other words, if we have one module with average occupancy of 500 preinitialized slots out of 1000, and 500 others with 1 slot each, and a new module comes along and wants a slot, we have only 1/501 chance of picking one of the 500.

To unbias this I should probably keep a global freelist (whole pool of choices mixed together), randomly pick from that freelist, and keep a reverse-index of slot to last-allocated module to remove the index from that module's freelist (or lazily do so next time we look at that list). I'll take a closer look at this tomorrow!

fitzgen commented on crates/runtime/src/module_id.rs:
```diff
@@ -7,6 +7,13 @@ use std::sync::atomic::{AtomicU64, Ordering};
+    pub fn none() -> Self {
+        Self(u64::MAX)
+    }
+}
```
Why not change `CompiledModuleId` to be a newtype around `NonZeroU64` and then we can use plain `Option` and `None` without pessimizing size, instead of having our own custom sentinel that we can forget to check for?

fitzgen replied:
Also, this is an existing issue (introduced in the memfd PR?) that is maybe going to become worse with these changes, but we use a relaxed ordering in the `fetch_add` when allocating new `CompiledModuleId`s below -- it seems like this could be a problem on non-x86?

cfallin replied:
Re: relaxed ordering, that's fine -- all that means is that the atomic ID-allocation event is not synchronized with respect to any other loads/stores in the system (there are no fences). But the fetch-add itself is still an atomic action, with some linear ordering amongst the other accesses to the "next ID" atomic. That's sufficient for the invariants we need: we just need unique IDs for each module, we don't care what they are.

(Good idea wrt `NonZeroU64`; I'll make that change!)

fitzgen replied:
Ah yes, I didn't think about how the `fetch_add` itself was still atomic. Thanks!

cfallin commented:
I've updated this PR now to use a better data structure/algorithm design. It performs a fair random choice of victim slot when no slots with the desired affinity are available, and it has all O(1) updates -- somewhat tricky given the need to maintain two freelists (global and per-module) and remove from both. This is done by keeping Vecs and using swap_remove, and tracking a slot's position in each freelist in a separate reverse-index. Hopefully the comments make this a little more clear.

I've added a randomized test that counts ID-reuse (a little random simulation of sorts) and verifies a reasonable hit rate (at least twice what would be expected with random reuse) as well.

fitzgen commented:
> The policy tracks a freelist per "compiled module ID", and when
> allocating a slot for an instance, tries these three options in order:
>
> 1. A slot from the freelist for this module (i.e., last used for another
>   instantiation of this particular module), or
> 3. A slot that was last used by some other module or never before.

1..3 :eyes:

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+    ///
+    /// The data structures here deserve a little explanation:
+    ///
+    /// - free_list: this is a vec of slot indices that are free, no
+    ///   matter their affinities.
```
This is *all* slot indices that are free, right? That is, an invariant of this code is that every free slot index is always in this list, right?

If my understanding is correct, then I think it makes sense to clarify this comment as such.

cfallin replied:
Yep, clarified, thanks!

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+            _ => unreachable!(),
+        }
+    }
+
+    /// Update the index of this slot in the global free list.
+    fn update_free_list_index(&mut self, index: usize) {
+        match self {
+            &mut Self::Empty {
+                ref mut free_list_index,
+            }
+            | &mut Self::Affinity {
+                ref mut free_list_index,
+                ..
+            } => {
+                *free_list_index = index;
+            }
+            _ => panic!("Taken in free list"),
+        }
+    }
+
+    /// Get the index of this slot in its per-module free list.
+    fn per_module_index(&self) -> usize {
+        match self {
+            &Self::Affinity {
+                per_module_index, ..
+            } => per_module_index,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Update the index of this slot in its per-module free list.
+    fn update_per_module_index(&mut self, index: usize) {
+        match self {
+            &mut Self::Affinity {
+                ref mut per_module_index,
+                ..
+            } => {
+                *per_module_index = index;
+            }
+            _ => panic!("Taken in per-module free list"),
```
All of the methods on `SlotState` panic if the `SlotState` is `Taken`. What do you think of refactoring `SlotState` to look like this:

```rust
enum SlotState {
    Taken,
    Free(FreeSlotState),
}

impl SlotState {
    // Do the expected matching where `Taken` returns `None` here...
    fn as_free_slot(&self) -> Option<&FreeSlotState> { ... }
    fn as_free_slot_mut(&mut self) -> Option<&mut FreeSlotState> { ... }
}

enum FreeSlotState {
    Empty { ... },
    Affinity { ... },
}

impl FreeSlotState {
    // all the methods above, which don't need to have panicking paths anymore
}
```

This way we can consolidate all the is-this-a-taken-slot checking into a single `slot_state[i].as_free_slot().expect("blah blah blah")` call, our types give us guarantees about their data, and potential panics aren't hidden behind method calls.

fitzgen replied:
Could also have `SlotState::unwrap_free_slot[_mut]` methods to avoid the `exepect` at every call site, but still have the potential panics "visible" to callers.

cfallin replied:
Sure, I refactored along these lines (and took the route of panic'ing in the as_free accessors). Good idea, thanks!

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+        }
+    }
+
+    /// Internal: remove a slot-index from a per-module free list.
+    fn remove_module_free_list_item(
+        slot_state: &mut Vec<SlotState>,
+        per_module: &mut HashMap<CompiledModuleId, Vec<usize>>,
+        id: CompiledModuleId,
+        index: usize,
+    ) {
+        let per_module_list = per_module.get_mut(&id).unwrap();
+        let per_module_index = slot_state[index].per_module_index();
+        assert_eq!(index, per_module_list.swap_remove(per_module_index));
+        if per_module_index < per_module_list.len() {
+            let replaced = per_module_list[per_module_index];
+            slot_state[replaced].update_per_module_index(per_module_index);
+        }
+        if per_module_list.is_empty() {
+            per_module.remove(&id);
+        }
+    }
```
Nitpick: sort of feels like this should be a free function.

cfallin replied:
Done!

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+            let replaced = per_module_list[per_module_index];
+            slot_state[replaced].update_per_module_index(per_module_index);
+        }
+        if per_module_list.is_empty() {
+            per_module.remove(&id);
+        }
```
Maybe

```rust
debug_assert!(per_module.contains(&id));
debug_assert!(!per_module.get(&id).is_empty());
```

at the start of this function?

cfallin replied:
The first is covered by the `get_mut()` at the top; I added a `.expect()` rather than `.unwrap()` to make any failure clearer though. Added the second.

fitzgen replied:
> The first is covered by the `get_mut()` at the top

FWIW, I don't mind having "redundant" assertions because, in some sense, debug assertions are *always* redundant, since we are checking things that we believe should always be true.

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+        // we don't have an appropriate slot with the preferred
+        // affinity.
+        free_list: Vec<usize>,
+        // Invariant: any module ID in this hashmap must have a
+        // non-empty list of free slots (otherwise we remove it).
+        per_module: HashMap<CompiledModuleId, Vec<usize>>,
```
Is the idea that as `Module`s get dropped, we don't need to actively empty the `per_module` free lists, because they will eventually get stolen by other instance allocations, and once the dropped module's free list is empty, it is removed from the `per_module` free lists? I think this would be good to test somehow, since it seems like an easy memory leak if we mess this up. Also, it seems like it would be good to document in a comment somewhere.

cfallin replied:
Yep, exactly; a slot's affinity remains until the slot is overwritten (reused for a different module). Extended the doc-comment, and added some testing for the keys of this hashmap (and that its values are non-empty lists) as well.

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+    }
+
+    /// Internal: remove a slot-index from the global free list.
+    fn remove_free_list_item(
```
Nitpick: maybe name this `remove_global_free_list_item` to clarify the difference between this and `remove_module_free_list_item`?

Ditto about freeling like a free function.

cfallin replied:
Yup, done on both.

cfallin commented:
> 1..3 :eyes:

Incomplete edit, sorry! The distinction between the last two (empty, and then slot with other affinity) was removed because it made the data structure simpler, and in steady-state (past the first `n_slot` instantiations in the process) no slots will be empty.

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+                    // the requested module-ID. Pick the last one; any
+                    // will do, no need for randomness here.
+                    assert!(!this_module.is_empty());
+                    let new_id = this_module.pop().expect("List should never be empty");
```
It would be nice to have a newtype for slot indices (`SlotId`?) too.

`new_id` threw me for a little bit of a loop, and made me double check that this wasn't a `CompiledModuleId` and was actually a slot index, which is what I expected it to be, but just the naming made me doubt myself enough that I had to double check.

fitzgen replied:
and/or we could `s/new_id/slot_index/` here

cfallin replied:
I newtyped All The Things (`SlotId`, `GlobalFreeListIndex`, `PerModuleFreeListIndex`); hopefully this is better!

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+                    // random choice among modules, which is just as
+                    // complex as this process.
+                    //
+                    // We don't bother picking an empty slot (no
+                    // established affinity) before a random slot,
+                    // because this is more complex, and in the steady
+                    // state, all slots will see at least one
+                    // instantiation very quickly, so there will never
+                    // (past an initial phase) be a slot with no
+                    // affinity.
```
This is a very helpful comment, thanks for writing this!

cfallin replied:
:+1:

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+        }
+    }
+
+    pub(crate) fn free(&mut self, index: usize, id: Option<CompiledModuleId>) {
```
I think the `id` is unnecessary here, and we can recover this from the `slot_state[index]`, if it exists? If so, removing `id` would help move us towards a single source of truth with fewer things that can get misaligned and out of sync, and would allow us to clean up slots that have a module affinity even when callers forgot to pass a module id here.

cfallin replied:
Yes, that's a good point; made it a field of the `Taken` state now, to remember it for the free.

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+                let free_list_index = free_list.len();
+                free_list.push(index);
+                if let Some(id) = id {
+                    let per_module_list = per_module.entry(id).or_insert_with(|| vec![]);
```
Nitpick: we know we are going to insert into this new vec so we can do `.or_insert_with(|| Vec::with_capacity(1))` here.

cfallin replied:
Good point!

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -0,0 +1,408 @@
+                        per_module_index,
+                    };
+                } else {
+                    slot_state[index] = SlotState::Empty { free_list_index };
```
Yeah so calling `free` without an `id` will create new `SlotState::Empty`s which goes against that steady-state comment above (ignoring a global analysis of all `free` callers). So yeah I think removing the `id` param would be nice if we can, since it also helps make it easier to verify statements about the correctness of this system.

cfallin replied:
Indeed. The intent was to be a little more general (one could in theory choose *not* to retain the state that makes a slot have affinity) but as you say that complexifies things a bit and we should probably approach that as we need it. (And it need not necessarily touch the affinity state either; there's little harm in a stray affinity for a module that will never exist again.)

fitzgen reviewed (commented):
This is looking really good! A couple questions, suggestions, and nitpicks below.

cfallin commented:
I think I addressed all your comments; thanks @fitzgen ! This is rebased on the latest #3697 as well.

fitzgen commented on crates/runtime/src/instance/allocator/pooling/index_allocator.rs:
```diff
@@ -30,56 +61,112 @@ pub(crate) enum PoolingAllocationState {
+}
+
+impl SlotState {
+    fn as_free(&self) -> &FreeSlotState {
```
I'd personally add `unwrap_` as a prefix to these match-on-a-variant-and-return-the-inner-bits-or-else-panic methods. Typically a plain `as_blah` would return an `Option<&T>` if it is possibly not a `T`, rather than panicking.

cfallin replied:
Done!

fitzgen reviewed (approved):
Thanks! (Note that I haven't looked at any of the earlier commits)
